File: explore/inat_auth.py
"""
inat_auth.py
~~~~~~~~~~~~
Credential storage and token resolution for iNaturalist.

Design goals
------------
* Keep secrets **out of the repo**. Nothing here ever writes a credential to a
  file inside the working tree.
* Store secrets in the OS credential vault (Windows Credential Manager /
  macOS Keychain / Secret Service) via ``keyring``.  This mirrors what the
  Lightroom plugin will do with the ``LrPasswords`` SDK namespace, which is
  backed by the same OS vaults -- so the *shape* of the credential model
  transfers directly even though the two runtimes cannot share a vault entry
  namespace.
* Support more than one auth mode, because iNaturalist offers several and the
  right one for the shipping plugin is still an open question.

Auth modes
----------
``api_token``
    A JSON Web Token copied from https://www.inaturalist.org/users/api_token
    while signed in to the website.  No OAuth application registration
    required.  Short-lived (~24 h), so it is only suitable for exploration --
    but it unblocks testing today.

``oauth_password``
    OAuth 2.0 Resource Owner Password Credentials grant.  Needs a registered
    application (client id + secret) plus the account username and password.
    Yields a long-lived OAuth token, which is then exchanged for a JWT.

Both modes converge on the same thing: a bearer token accepted by
``https://api.inaturalist.org/v1``.

CLI
---
    python inat_auth.py status
    python inat_auth.py store-token            # prompts for a pasted JWT
    python inat_auth.py store-oauth-app        # prompts for app id/secret/user/pass
    python inat_auth.py clear
"""

from __future__ import annotations

import argparse
import getpass
import json
import os
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import requests
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent / ".env")

# --------------------------------------------------------------------------
# Constants
# --------------------------------------------------------------------------

WWW_BASE = "https://www.inaturalist.org"
API_V1 = "https://api.inaturalist.org/v1"

#: Vault "service" name.  Chosen to match the Lightroom plugin's intended
#: LrPasswords key prefix so the two stay conceptually aligned.
KEYRING_SERVICE = "inat-lightroom"

#: Individual entries stored under KEYRING_SERVICE.
_KEY_API_TOKEN = "api_token"
_KEY_API_TOKEN_AT = "api_token_obtained_at"
_KEY_APP_ID = "app_id"
_KEY_APP_SECRET = "app_secret"
_KEY_USERNAME = "username"
_KEY_PASSWORD = "user_password"

#: iNaturalist JWTs are documented as valid for 24 hours.  Refresh a little
#: early so a long-running script does not expire mid-flight.
JWT_LIFETIME_SECONDS = 24 * 60 * 60
JWT_REFRESH_MARGIN_SECONDS = 60 * 60

USER_AGENT = "inat-lightroom-explore/0.1 (+https://github.com/WilliamsJason/inat-lightroom)"


class AuthError(RuntimeError):
    """Raised when a usable token cannot be produced."""


# --------------------------------------------------------------------------
# Vault access
# --------------------------------------------------------------------------


def _keyring():
    """Import keyring lazily so the module is importable without a vault."""
    try:
        import keyring

        return keyring
    except Exception as exc:  # pragma: no cover - environment dependent
        raise AuthError(f"keyring is unavailable: {exc}") from exc


def vault_get(key: str) -> str | None:
    try:
        return _keyring().get_password(KEYRING_SERVICE, key)
    except Exception:
        return None


def vault_set(key: str, value: str) -> None:
    _keyring().set_password(KEYRING_SERVICE, key, value)


def vault_delete(key: str) -> None:
    kr = _keyring()
    try:
        kr.delete_password(KEYRING_SERVICE, key)
    except Exception:
        pass


def _setting(env_name: str, vault_key: str) -> str | None:
    """Environment variable wins over the vault, so CI/one-offs can override."""
    return os.environ.get(env_name) or vault_get(vault_key)


# --------------------------------------------------------------------------
# Credential model
# --------------------------------------------------------------------------


@dataclass
class OAuthAppCredentials:
    app_id: str
    app_secret: str
    username: str
    password: str


def load_oauth_app_credentials() -> OAuthAppCredentials | None:
    app_id = _setting("INAT_APP_ID", _KEY_APP_ID)
    app_secret = _setting("INAT_APP_SECRET", _KEY_APP_SECRET)
    username = _setting("INAT_USERNAME", _KEY_USERNAME)
    password = _setting("INAT_PASSWORD", _KEY_PASSWORD)
    if all([app_id, app_secret, username, password]):
        return OAuthAppCredentials(app_id, app_secret, username, password)  # type: ignore[arg-type]
    return None


def store_api_token(token: str) -> None:
    vault_set(_KEY_API_TOKEN, token.strip())
    vault_set(_KEY_API_TOKEN_AT, str(int(time.time())))


def store_oauth_app_credentials(creds: OAuthAppCredentials) -> None:
    vault_set(_KEY_APP_ID, creds.app_id)
    vault_set(_KEY_APP_SECRET, creds.app_secret)
    vault_set(_KEY_USERNAME, creds.username)
    vault_set(_KEY_PASSWORD, creds.password)


def clear_all() -> None:
    for key in (
        _KEY_API_TOKEN,
        _KEY_API_TOKEN_AT,
        _KEY_APP_ID,
        _KEY_APP_SECRET,
        _KEY_USERNAME,
        _KEY_PASSWORD,
    ):
        vault_delete(key)


# --------------------------------------------------------------------------
# Token acquisition
# --------------------------------------------------------------------------


def _cached_api_token() -> str | None:
    token = _setting("INAT_API_TOKEN", _KEY_API_TOKEN)
    if not token:
        return None
    # An env-supplied token has no recorded age; assume the caller knows.
    if os.environ.get("INAT_API_TOKEN"):
        return token
    obtained_at = vault_get(_KEY_API_TOKEN_AT)
    if obtained_at:
        age = time.time() - int(obtained_at)
        if age > JWT_LIFETIME_SECONDS - JWT_REFRESH_MARGIN_SECONDS:
            return None
    return token


def _oauth_password_grant(creds: OAuthAppCredentials) -> str:
    """Exchange username/password for a long-lived OAuth access token."""
    payload = {
        "grant_type": "password",
        "client_id": creds.app_id,
        "client_secret": creds.app_secret,
        "username": creds.username,
        "password": creds.password,
    }
    resp = requests.post(
        f"{WWW_BASE}/oauth/token",
        json=payload,
        headers={"User-Agent": USER_AGENT},
        timeout=30,
    )
    if resp.status_code >= 400:
        raise AuthError(
            f"OAuth token request failed ({resp.status_code}): {resp.text[:400]}"
        )
    token = resp.json().get("access_token")
    if not token:
        raise AuthError(f"No access_token in OAuth response: {resp.text[:400]}")
    return token


def _exchange_for_jwt(oauth_token: str) -> str:
    """Trade an OAuth bearer token for the JWT used by the modern API."""
    resp = requests.get(
        f"{WWW_BASE}/users/api_token",
        headers={"Authorization": f"Bearer {oauth_token}", "User-Agent": USER_AGENT},
        timeout=30,
    )
    if resp.status_code >= 400:
        raise AuthError(
            f"JWT exchange failed ({resp.status_code}): {resp.text[:400]}"
        )
    token = resp.json().get("api_token")
    if not token:
        raise AuthError(f"No api_token in response: {resp.text[:400]}")
    return token


def get_token(*, refresh: bool = False) -> str:
    """
    Return a bearer token usable against the iNaturalist v1 API.

    Resolution order:
      1. ``INAT_API_TOKEN`` environment variable
      2. A non-expired JWT in the OS credential vault
      3. OAuth password grant using stored/env application credentials,
         exchanged for a fresh JWT (which is then cached)
    """
    if not refresh:
        cached = _cached_api_token()
        if cached:
            return cached

    creds = load_oauth_app_credentials()
    if creds is None:
        had_expired_token = bool(_setting("INAT_API_TOKEN", _KEY_API_TOKEN))
        if had_expired_token:
            raise AuthError(
                "The stored iNaturalist JWT has expired (they last 24 hours).\n"
                "Get a fresh one while signed in at\n"
                "  https://www.inaturalist.org/users/api_token\n"
                "then run `python inat_auth.py store-token` again.\n\n"
                "To stop doing this by hand you need an approved OAuth "
                "application; see explore/README.md."
            )
        raise AuthError(
            "No usable iNaturalist credentials found.\n"
            "Either:\n"
            "  * run `python inat_auth.py store-token` and paste the JWT from\n"
            "    https://www.inaturalist.org/users/api_token, or\n"
            "  * run `python inat_auth.py store-oauth-app` to save an approved\n"
            "    OAuth application id/secret plus your username and password."
        )

    jwt = _exchange_for_jwt(_oauth_password_grant(creds))
    store_api_token(jwt)
    return jwt


def auth_headers(token: str | None = None) -> dict[str, str]:
    """Standard headers for an authenticated API call."""
    return {
        "Authorization": f"Bearer {token or get_token()}",
        "User-Agent": USER_AGENT,
    }


def whoami(token: str | None = None) -> dict[str, Any]:
    """Verify a token by fetching the authenticated user."""
    resp = requests.get(
        f"{API_V1}/users/me", headers=auth_headers(token), timeout=30
    )
    if resp.status_code >= 400:
        raise AuthError(f"Token check failed ({resp.status_code}): {resp.text[:400]}")
    results = resp.json().get("results") or []
    if not results:
        raise AuthError("users/me returned no results")
    return results[0]


# --------------------------------------------------------------------------
# CLI
# --------------------------------------------------------------------------


def _cmd_status(_args: argparse.Namespace) -> int:
    have_token = bool(_cached_api_token())
    have_app = load_oauth_app_credentials() is not None
    print(f"Cached JWT present : {have_token}")
    print(f"OAuth app creds    : {have_app}")
    if not (have_token or have_app):
        print("\nNo credentials stored. Run `store-token` or `store-oauth-app`.")
        return 1
    try:
        user = whoami()
    except AuthError as exc:
        print(f"\nToken check FAILED: {exc}")
        return 1
    print(f"\nAuthenticated as   : {user.get('login')} (id={user.get('id')})")
    print(f"Observations       : {user.get('observations_count')}")
    return 0


def _cmd_store_token(_args: argparse.Namespace) -> int:
    print("Sign in at https://www.inaturalist.org then open:")
    print("  https://www.inaturalist.org/users/api_token")
    print("Paste either the raw token or the whole JSON blob.\n")
    raw = getpass.getpass("api_token: ").strip()
    if not raw:
        print("Nothing entered.", file=sys.stderr)
        return 1
    # Accept a pasted {"api_token": "..."} payload as well as a bare token.
    if raw.startswith("{"):
        try:
            raw = json.loads(raw)["api_token"]
        except Exception as exc:
            print(f"Could not parse JSON: {exc}", file=sys.stderr)
            return 1
    store_api_token(raw)
    try:
        user = whoami(raw)
    except AuthError as exc:
        print(f"Stored, but the token did not validate: {exc}", file=sys.stderr)
        return 1
    print(f"Stored. Authenticated as {user.get('login')} (id={user.get('id')}).")
    return 0


def _cmd_store_oauth_app(args: argparse.Namespace) -> int:
    if getattr(args, "from_env", False):
        creds = load_oauth_app_credentials()
        if creds is None:
            print(
                "Could not find all four values in the environment or .env file.\n"
                "Expected INAT_APP_ID, INAT_APP_SECRET, INAT_USERNAME, INAT_PASSWORD.",
                file=sys.stderr,
            )
            return 1
        print("Migrating credentials from the environment into the OS vault.")
    else:
        print(
            "Create an application at https://www.inaturalist.org/oauth/applications/new"
        )
        print("if you have not already.\n")
        app_id = input("Application client ID: ").strip()
        app_secret = getpass.getpass("Application client secret: ").strip()
        username = input("iNaturalist username: ").strip()
        password = getpass.getpass("iNaturalist password: ").strip()
        if not all([app_id, app_secret, username, password]):
            print("All four values are required.", file=sys.stderr)
            return 1
        creds = OAuthAppCredentials(app_id, app_secret, username, password)

    store_oauth_app_credentials(creds)
    try:
        token = get_token(refresh=True)
        user = whoami(token)
    except AuthError as exc:
        print(f"Stored, but authentication failed: {exc}", file=sys.stderr)
        return 1
    print(f"Stored. Authenticated as {user.get('login')} (id={user.get('id')}).")
    if getattr(args, "from_env", False):
        env_file = Path(__file__).parent / ".env"
        if env_file.exists():
            print(
                f"\nCredentials now live in the OS vault. You can safely delete\n"
                f"  {env_file}"
            )
    return 0


def _cmd_clear(_args: argparse.Namespace) -> int:
    clear_all()
    print("Cleared all stored iNaturalist credentials.")
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Manage iNaturalist credentials.")
    sub = parser.add_subparsers(dest="command", required=True)

    sub.add_parser("status", help="Show what is stored and validate it.").set_defaults(
        func=_cmd_status
    )
    sub.add_parser("store-token", help="Store a pasted JWT API token.").set_defaults(
        func=_cmd_store_token
    )
    store_app = sub.add_parser(
        "store-oauth-app", help="Store OAuth application + account credentials."
    )
    store_app.add_argument(
        "--from-env",
        action="store_true",
        help="Read the four values from the environment / .env instead of prompting, "
        "then move them into the OS vault.",
    )
    store_app.set_defaults(func=_cmd_store_oauth_app)
    sub.add_parser("clear", help="Remove all stored credentials.").set_defaults(
        func=_cmd_clear
    )

    args = parser.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    raise SystemExit(main())

File: plugin/pinned.lrplugin/InatAuth.lua
--[[
  InatAuth.lua
  ------------
  Token acquisition and credential storage for iNaturalist.

  There are two tokens involved, and conflating them is the easiest way to
  break this plugin:

    * OAuth access token -- never expires, but the v1 API does NOT accept it
      for writes. Its only real job is to be exchanged for a JWT.
    * JWT "API token"    -- expires after 24 hours, and is what actually
      authenticates API calls.

  Presenting a bare OAuth token to api.inaturalist.org does not return 401.
  The request is silently processed as an anonymous user, so writes fail in
  ways that are very hard to diagnose. Always send the JWT.

  One way to obtain one today: the user signs in at inaturalist.org, opens
  /users/api_token, and pastes the JWT into the setup dialog. It must be
  repeated every 24 hours. It stays because it needs no registered
  application, which makes it the fallback when anything about OAuth is
  unavailable.

  The better way, and now the default: the authorization code flow with PKCE,
  in InatOAuth.lua. The user signs in once in their browser, the plugin keeps
  a never-expiring OAuth access token, and this module quietly mints a new JWT
  from it whenever the old one ages out. Nobody pastes anything again.

  This module used to offer a third way -- an OAuth application using the
  password grant, which exchanged the user's iNaturalist username and password
  for a never-expiring access token and minted JWTs from it. It worked, and it
  is gone anyway. iNaturalist recommends against the password grant and
  specifically against it in publicly distributed applications, because it
  requires typing an account password into third-party software. The advice is
  the authorization code flow with PKCE, which sends the user to iNaturalist to
  sign in and never lets the plugin near the password.

  It was also impractical: the app id and secret are per-application, and
  iNaturalist reviews applications by hand, so every user would have needed
  their own approved application before the fields did anything. PKCE does not
  have that problem -- one approved application, no secret, and every user
  authenticating against it as themselves.

  URLHandler.lua and PluginUrls.lua receive the redirect that carries the
  authorization code back; that is what Info.lua's URLHandler entry is for.

  Secrets are held by LrPasswords, which is backed by the OS credential
  vault. Non-secret bookkeeping (when the JWT was obtained, which mode is in
  use) lives in LrPrefs.
--]]

local LrDialogs = import "LrDialogs"
local LrHttp    = import "LrHttp"
local LrPasswords = import "LrPasswords"
local LrPrefs   = import "LrPrefs"
local LrStringUtils = import "LrStringUtils"

local json = require "json"

local logger = require "Log"
local prefs  = LrPrefs.prefsForPlugin()

local WWW_BASE = "https://www.inaturalist.org"
local API_V1   = "https://api.inaturalist.org/v1"

-- LrPasswords keys. LrPasswords is already scoped to this plugin, so these
-- do not need the toolkit identifier prefixed.
local KEY_API_TOKEN    = "api_token"
local KEY_OAUTH_TOKEN  = "oauth_access_token"


-- iNaturalist JWTs last 24 hours. Refresh early so a long export does not
-- expire midway through.
local JWT_LIFETIME_SECONDS = 24 * 60 * 60
local JWT_REFRESH_MARGIN   = 60 * 60

local InatAuth = {}

--------------------------------------------------------------------------------
-- Credential storage
--------------------------------------------------------------------------------

--- Read a secret, normalising the empty string to nil.
local function retrieve(key)
  local value = LrPasswords.retrieve(key)
  if value == nil or value == "" then
    return nil
  end
  return value
end

local function store(key, value)
  LrPasswords.store(key, value or "")
end

--- Read the expiry out of a JWT.
--
-- A JWT is three base64url segments separated by dots, the middle one being
-- JSON with an "exp" claim holding a Unix timestamp. Reading it means expiry
-- is known rather than inferred from when the user happened to paste, which
-- matters when a token was copied some time before it was pasted in.
--
-- @return expiry as a Unix timestamp, or nil if it cannot be determined
local function decodeExpiry(token)
  local payload = token:match("^[^%.]+%.([^%.]+)%.")
  if not payload then return nil end

  -- base64url differs from base64 in two characters, and drops the padding.
  payload = payload:gsub("%-", "+"):gsub("_", "/")

  local remainder = #payload % 4
  if remainder == 2 then
    payload = payload .. "=="
  elseif remainder == 3 then
    payload = payload .. "="
  elseif remainder == 1 then
    return nil
  end

  local ok, decoded = pcall(LrStringUtils.decodeBase64, payload)
  if not ok or not decoded then return nil end

  local parsedOk, parsed = pcall(json.decode, decoded)
  if not parsedOk or type(parsed) ~= "table" then return nil end

  return tonumber(parsed.exp)
end

--- Store a JWT pasted by the user.
-- Accepts either a bare token or the whole {"api_token":"..."} response body.
function InatAuth.storeApiToken(raw)
  if not raw or raw == "" then
    return false, "No token supplied."
  end

  local token = raw:gsub("^%s+", ""):gsub("%s+$", "")

  if token:sub(1, 1) == "{" then
    local ok, decoded = pcall(json.decode, token)
    if not ok or type(decoded) ~= "table" or not decoded.api_token then
      return false, "That looks like JSON but has no api_token field."
    end
    token = decoded.api_token
  end

  -- Catch the common paste mistakes -- grabbing the page text, the URL, or
  -- only part of the token -- before they turn into a confusing API error.
  if not token:match("^[%w%-_]+%.[%w%-_]+%.[%w%-_]+$") then
    return false, "That does not look like an iNaturalist API token.\n\n"
      .. "Expected three dot-separated blocks. Copy the value of the "
      .. "\"api_token\" field from www.inaturalist.org/users/api_token."
  end

  local expiresAt = decodeExpiry(token)
  if expiresAt and expiresAt <= os.time() then
    return false, "That token has already expired.\n\n"
      .. "Reload www.inaturalist.org/users/api_token to get a fresh one."
  end

  store(KEY_API_TOKEN, token)
  prefs.apiTokenObtainedAt = os.time()
  prefs.apiTokenExpiresAt  = expiresAt
  prefs.authMode = "manual_jwt"
  logger:info("Stored a manually pasted JWT"
    .. (expiresAt and (", expires at " .. tostring(expiresAt)) or ""))
  return true, nil
end

function InatAuth.clear()
  store(KEY_API_TOKEN, "")
  store(KEY_OAUTH_TOKEN, "")
  prefs.apiTokenObtainedAt = nil
  prefs.apiTokenExpiresAt  = nil
  prefs.authMode = nil

  -- Any half-finished browser sign-in goes too. Clearing credentials is what
  -- someone does when they want the plugin to have forgotten them, and a
  -- verifier sitting in the vault waiting for a redirect is not that.
  local ok, InatOAuth = pcall(require, "InatOAuth")
  if ok and InatOAuth then
    pcall(InatOAuth.clearPending)
  end

  logger:info("Cleared stored credentials")
end

--------------------------------------------------------------------------------
-- OAuth
--------------------------------------------------------------------------------

--- Store the never-expiring OAuth access token from a browser sign-in.
--
-- This is not the token the API accepts. Its whole job is to sit in the vault
-- and mint 24-hour JWTs on demand, which is what makes browser sign-in a
-- one-time act rather than a daily one.
--
-- @return true, or false plus a reason
function InatAuth.storeOAuthToken(token)
  if type(token) ~= "string" or token == "" then
    return false, "iNaturalist returned an empty access token."
  end

  store(KEY_OAUTH_TOKEN, token)
  -- Any previously pasted JWT is now stale bookkeeping: the next getToken
  -- should mint a fresh one rather than trust what the old mode left behind.
  store(KEY_API_TOKEN, "")
  prefs.apiTokenObtainedAt = nil
  prefs.apiTokenExpiresAt  = nil
  prefs.authMode = "oauth"

  logger:info("Stored an OAuth access token")
  return true, nil
end

--- Whether the plugin is signed in through the browser.
function InatAuth.isSignedIn()
  return retrieve(KEY_OAUTH_TOKEN) ~= nil
end

--- Exchange the stored OAuth token for a fresh 24-hour JWT.
--
-- Must be called from inside a task: LrHttp.get yields.
--
-- The endpoint is on www.inaturalist.org rather than the API host, and it is
-- the one place a bare OAuth bearer token is the right thing to send.
--
-- @return jwt string, or nil plus a reason
local function refreshJwtFromOAuth()
  local oauthToken = retrieve(KEY_OAUTH_TOKEN)
  if not oauthToken then
    return nil, "Not signed in to iNaturalist."
  end

  local headers = {
    { field = "Authorization", value = "Bearer " .. oauthToken },
    { field = "Accept",        value = "application/json" },
  }

  local respBody, respHeaders = LrHttp.get(WWW_BASE .. "/users/api_token", headers)
  if not respBody then
    return nil, "Could not reach iNaturalist to refresh the API token."
  end

  local status = respHeaders and tonumber(respHeaders.status)
  if status and status >= 400 then
    -- 401 here means the access token has been revoked -- from iNaturalist's
    -- account settings, or by the application being deleted. There is no way
    -- back from the plugin's side, so say what actually fixes it.
    if status == 401 then
      return nil, "iNaturalist has rejected this plugin's access.\n\n"
        .. "That usually means access was revoked on the website. Open\n"
        .. "File > Plug-in Extras > Pinned Settings… and sign in again."
    end
    return nil, "iNaturalist refused to issue an API token (HTTP "
      .. tostring(status) .. ")."
  end

  local ok, data = pcall(json.decode, respBody)
  if not ok or type(data) ~= "table" or not data.api_token then
    return nil, "iNaturalist's reply contained no API token."
  end

  local token = data.api_token
  store(KEY_API_TOKEN, token)
  prefs.apiTokenObtainedAt = os.time()
  prefs.apiTokenExpiresAt  = decodeExpiry(token)
  prefs.authMode = "oauth"

  logger:info("Refreshed the API token from the stored OAuth token")
  return token, nil
end

--- Seconds since the stored JWT was obtained; nil when none is stored.
function InatAuth.tokenAgeSeconds()
  if not prefs.apiTokenObtainedAt then return nil end
  return os.time() - prefs.apiTokenObtainedAt
end

--- Seconds until the stored JWT expires. Negative when already expired,
-- nil when there is no token or its expiry could not be determined.
function InatAuth.tokenSecondsRemaining()
  if prefs.apiTokenExpiresAt then
    return prefs.apiTokenExpiresAt - os.time()
  end

  -- No decoded expiry: fall back to assuming the full lifetime from when it
  -- was stored.
  if prefs.apiTokenObtainedAt then
    return (prefs.apiTokenObtainedAt + JWT_LIFETIME_SECONDS) - os.time()
  end

  return nil
end

--- Return the stored JWT if it is still usable, else nil.
local function cachedTokenIfUsable()
  local token = retrieve(KEY_API_TOKEN)
  if not token then return nil end

  local remaining = InatAuth.tokenSecondsRemaining()
  if not remaining then return nil end

  -- Refresh early so a long export cannot have its token die mid-run.
  if remaining <= JWT_REFRESH_MARGIN then return nil end

  return token
end

--------------------------------------------------------------------------------
-- Public token accessor
--------------------------------------------------------------------------------

--- Return a JWT suitable for the v1 API.
--
-- Two ways to have one, and they behave differently:
--
--   * Signed in through the browser. The stored OAuth token never expires, so
--     an expired JWT is refreshed silently and the user is never asked for
--     anything. This is the branch the module was always written around.
--   * A pasted token. There is nothing to refresh from, so an expired one can
--     only be reported.
--
-- Must be called from inside an async task, because LrHttp yields.
--
-- @param forceRefresh  Mint a new JWT even if the cached one looks usable.
--                      Honoured only when signed in through the browser: with
--                      a pasted token there is nothing to refresh from, and
--                      honouring it there is what previously made a token that
--                      had just been pasted report itself as expired. Callers
--                      pass it after saving credentials, so it has to be
--                      harmless rather than an error.
-- @return token string, or nil plus an error message
function InatAuth.getToken(forceRefresh)
  local signedIn = InatAuth.isSignedIn()

  if not (forceRefresh and signedIn) then
    local cached = cachedTokenIfUsable()
    if cached then
      return cached, nil
    end
  end

  if signedIn then
    return refreshJwtFromOAuth()
  end

  if retrieve(KEY_API_TOKEN) then
    return nil, "Your iNaturalist token has expired. Tokens last 24 hours.\n\n"
      .. "Sign in at inaturalist.org, open www.inaturalist.org/users/api_token, "
      .. "and paste the new token via\n"
      .. "File > Plug-in Extras > Pinned Settings…\n\n"
      .. "Or use Sign In with iNaturalist there once, and this stops happening."
  end

  return nil, "iNaturalist credentials are not set up.\n\n"
    .. "Use File > Plug-in Extras > Pinned Settings…."
end

--- A one-line summary of why a token could not be had.
--
-- Read from the stored state rather than from the caller's message, because
-- the messages are written for a dialog that is no longer shown and end with
-- directions to the very window that is about to open.
local function setupNotice()
  if retrieve(KEY_API_TOKEN) then
    return "Pinned's iNaturalist token has expired."
  end
  return "Pinned needs your iNaturalist credentials before it can do that."
end

--- Whether the settings window is the answer to this failure.
--
-- It is when there is nothing stored that the plugin can renew on its own: no
-- credentials at all, or a pasted token that has run out. It is not when the
-- user is signed in through the browser, because then the plugin renews its
-- own token and a failure is the network, a revocation, or iNaturalist being
-- down -- none of which are fixed by the window, and the first of which comes
-- and goes. Yanking the settings open over a dropped connection teaches the
-- user to close it without reading.
local function setupWouldHelp()
  return not InatAuth.isSignedIn()
end

--- Send the user to the place that fixes missing or expired credentials.
--
-- One route for every feature that needs a token, because it is one problem
-- with one fix, and it is not about the feature that happened to hit it. Four
-- call sites had drifted into three different titles and two severities, so
-- the same sentence looked like a different fault depending on which button
-- had been pressed.
--
-- It used to be a warning dialog whose whole content was directions to
-- File > Plug-in Extras > Pinned Settings…. Reading out a menu path is a
-- worse version of opening the window, so when the window is the answer it
-- opens the window -- on the Account tab, with the reason above the fields
-- that answer it. The caller's message is kept for the log, which is where
-- the detail is still wanted, and is still shown when the window is not the
-- answer or cannot be opened.
--
-- Required lazily: SettingsDialog requires this module at load time, so a
-- require at the top of the file would be circular.
function InatAuth.reportMissingCredentials(message)
  local fallback = message or "iNaturalist credentials are not set up.\n\n"
    .. "Use File > Plug-in Extras > Pinned Settings…."

  logger:info("No usable token: " .. tostring(fallback))

  local ok, opened = pcall(function()
    if not setupWouldHelp() then return false end

    local SettingsDialog = require "SettingsDialog"

    -- Already there. Opening a second copy of the window on top of the one
    -- the user is looking at would hide the fields that fix this.
    if SettingsDialog.isShowing() then return false end

    SettingsDialog.show({ tab = "account", notice = setupNotice() })
    return true
  end)

  if not ok then
    logger:error("Could not open settings: " .. tostring(opened))
  end

  -- Either the window was the wrong answer, or opening it failed -- and a
  -- feature that stops with nothing shown at all reads as one that quietly
  -- did nothing.
  if not ok or not opened then
    LrDialogs.message("Pinned", fallback, "warning")
  end
end

--- Verify a token by fetching the authenticated user.
-- @return user table (login, id, ...), or nil plus an error message
function InatAuth.whoami(token)
  local headers = {
    { field = "Authorization", value = "Bearer " .. token },
  }

  local respBody, respHeaders = LrHttp.get(API_V1 .. "/users/me", headers)
  if not respBody then
    return nil, "No response from the API"
  end

  local status = respHeaders and tonumber(respHeaders.status)
  if status and status >= 400 then
    return nil, "Token rejected (HTTP " .. tostring(status) .. ")"
  end

  local ok, data = pcall(json.decode, respBody)
  if not ok or type(data) ~= "table" then
    return nil, "Could not parse the response"
  end

  local results = data.results
  if not results or not results[1] then
    return nil, "The token does not identify a user"
  end

  return results[1], nil
end

return InatAuth
